test/e2e/resource_generators.go

```diff
@@ -0,0 +1,387 @@
+				SKU:       &imageSKU,
+				Version:   &imageVersion,
+			},
+			OSDisk: infrav1.OSDisk{
+				DiskSizeGB: 30,
+				OSType:     "Linux",
+				ManagedDisk: infrav1.ManagedDisk{
+					StorageAccountType: "Premium_LRS",
+				},
+			},
+		},
+	}
+	bootstrapConfig := &bootstrapv1.KubeadmConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Spec: bootstrapv1.KubeadmConfigSpec{
+			Files: []bootstrapv1.File{
+				{
+					Owner:       "root:root",
+					Path:        "/etc/kubernetes/azure.json",
+					Permissions: "0644",
+					Content:     cloudConfig(clusterName),
+				},
+			},
+			InitConfiguration: &kubeadmv1beta1.InitConfiguration{},
+			JoinConfiguration: &kubeadmv1beta1.JoinConfiguration{},
+		},
+	}
+	registrationOptions := kubeadmv1beta1.NodeRegistrationOptions{
+		Name: `{{ ds.meta_data["local_hostname"] }}`,
+		KubeletExtraArgs: map[string]string{
+			"cloud-provider": "azure",
+			"cloud-config":   "/etc/kubernetes/azure.json",
+		},
+	}
+
+	if firstControlPlane {
+		cpInitConfiguration(bootstrapConfig, registrationOptions)
+	} else {
+		cpJoinConfiguration(bootstrapConfig, registrationOptions)
+	}
+
+	machine := &capiv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Labels: map[string]string{
+				capiv1.MachineControlPlaneLabelName: "true",
+				capiv1.MachineClusterLabelName:      clusterName,
+			},
+		},
+		Spec: capiv1.MachineSpec{
+			Bootstrap: capiv1.Bootstrap{
+				ConfigRef: &corev1.ObjectReference{
+					APIVersion: bootstrapv1.GroupVersion.String(),
+					Kind:       framework.TypeToKind(bootstrapConfig),
+					Namespace:  bootstrapConfig.GetNamespace(),
+					Name:       bootstrapConfig.GetName(),
+				},
+			},
+			InfrastructureRef: corev1.ObjectReference{
+				APIVersion: infrav1.GroupVersion.String(),
+				Kind:       framework.TypeToKind(infraMachine),
+				Namespace:  infraMachine.GetNamespace(),
+				Name:       infraMachine.GetName(),
+			},
+			Version: &framework.Flags.KubernetesVersion,
+		},
+	}
+	return framework.Node{
+		Machine:         machine,
+		InfraMachine:    infraMachine,
+		BootstrapConfig: bootstrapConfig,
+	}
+}
+
+func cloudConfig(cn string) string {
+	// TODO This is ugly
+	return fmt.Sprintf(`{
+    "cloud": "AzurePublicCloud",
+    "tenantId": "%s",
+    "subscriptionId": "%s",
+    "aadClientId": "%s",
+    "aadClientSecret": "%s",
+    "resourceGroup": "%s",
+    "securityGroupName": "%s-controlplane-nsg",
+    "location": "westus2",
+    "vmType": "standard",
+    "vnetName": "%s-vnet",
+    "vnetResourceGroup": "%s",
+    "subnetName": "%s-controlplane-subnet",
+    "routeTableName": "%s-node-routetable",
+    "userAssignedID": "%s",
+    "loadBalancerSku": "standard",
+    "maximumLoadBalancerRuleCount": 250,
+    "useManagedIdentityExtension": false,
+    "useInstanceMetadata": true
+}`, creds.TenantID, creds.SubscriptionID, creds.ClientID, creds.ClientSecret, cn, cn, cn, cn, cn, cn, cn)
```

Comment:
seems using `creds` here gives a compilation error since its defined in a different file 
```
go test ./...
# sigs.k8s.io/cluster-api-provider-azure/test/e2e
test/e2e/resource_generators.go:211:5: undefined: creds
```

Reply:
Also maybe to make it more readable we can put the value in a struct and serialize to json?